src/query-builder/SelectQueryBuilder.ts

```diff
@@ -1875,11 +1875,28 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
                 queryRunner,
             )
             this.expressionMap.queryEntity = false
-            const cacheId = this.expressionMap.cacheId
-            // Creates a new cacheId for the count query, or it will retreive the above query results
-            // and count will return 0.
-            this.expressionMap.cacheId = cacheId ? `${cacheId}-count` : cacheId
-            const count = await this.executeCountQuery(queryRunner)
+
+            let count
+            if (
+                ((this.expressionMap.take !== undefined &&
+                    entitiesAndRaw.entities.length < this.expressionMap.take) ||
+                    (this.expressionMap.limit !== undefined &&
+                        entitiesAndRaw.entities.length <
+                            this.expressionMap.limit)) &&
+                this.expressionMap.skip === undefined &&
+                this.expressionMap.offset === undefined
```

Comment:
There are some edge cases for `take`/`limit`/`skip`/`offset`: `0`, `null`, `undefined`.